src/parser/uniterm.rs

```diff
@@ -0,0 +1,134 @@
+use super::*;
+use crate::types::UnboundTypeVariableError;
+use error::ParseError;
+use utils::{build_record, FieldPathElem};
+
+use crate::{
+    match_sharedterm,
+    position::{RawSpan, TermPos},
+    term::{MergePriority, MetaValue, RecordAttrs, RichTerm, Term},
+    types::{AbsType, Types},
+};
+
+use std::convert::TryFrom;
+
+/// A record in the UniTerm syntax.
+#[derive(Clone)]
+pub struct UniRecord {
+    pub fields: Vec<(FieldPathElem, RichTerm)>,
+    pub tail: Option<(Types, TermPos)>,
+    pub attrs: RecordAttrs,
+    pub pos: TermPos,
+}
+
+/// Error indicating that a construct is not allowed when trying to interpret an `UniRecord` as a
+/// record type in a strict way. See [`into_type_strict`]. Hold the position of the illegal
+/// construct.
+pub struct InvalidRecordTypeError(pub TermPos);
+
+impl UniRecord {
+    /// Try to convert a `UniRecord` to a type. The strict part means that if the `UniRecord` must
+    /// be a plain record type, uniquely containing fields of the form `fields: Type`. Currently,
+    /// it doesn't support the field path syntax: `{foo.bar.baz : Type}.into_type_strict()` returns
+    /// an `Err`.
+    pub fn into_type_strict(self) -> Result<Types, InvalidRecordTypeError> {
+        let ty = self.fields.into_iter()
+            // Because we build row types as a linked list by folding on the original iterator, the
+            // order of identifiers is reversed. This not a big deal but it's less confusing to the
+            // user to print them in the original order for error reporting.
+            .rev()
+            .try_fold(
+                self.tail.map(|(tail, _)| tail).unwrap_or(Types(AbsType::RowEmpty())),
+                |acc, (path_elem, rt)| {
+                    match path_elem {
+                        FieldPathElem::Ident(id) => {
+                            match_sharedterm! {rt.term, with {
+                                Term::MetaValue(MetaValue {
+                                    doc: None,
+                                    types: Some(ctrt),
+                                    contracts,
+                                    priority: MergePriority::Normal,
+                                    value: None,
+                                    }) if contracts.is_empty() =>
+                                        Ok(Types(AbsType::RowExtend(id, Some(Box::new(ctrt.types)), Box::new(acc)))),
+                                }
+                                else {
+                                    // Position of identifiers must always be set at this stage
+                                    // (parsing)
+                                    let span_id = id.pos.unwrap();
+                                    let term_pos = rt.pos.into_opt().unwrap_or(span_id);
+                                    Err(InvalidRecordTypeError(TermPos::Original(RawSpan::fuse(span_id, term_pos).unwrap())))
+                                }
+                            }
```

Comment:
But with your suggestion, we are doing a copy in any case, even if the pattern doesn't match. While the `assert_matches!` first checks before doing the clone, doesn't it?

It is probably cheap 99.9% of the time, but just for the sake of pretending: if `rt.term` is a `List` of size `10000`, then in the first case we don't copy anything, while with your suggestion, we would copy the whole array, I think?

Reply:
That's not really what the `match_sharedterm` is used for. You would use it for cases where you still need access to the `SharedTerm` in case no match arm is selected, so that you avoid to duplicate a `SharedTerm` to another (maybe it's not very clear and I should clarify the documentation for the macro).

Here since you don't need it, the regular match is enough. I'm quite sure that being one of the classic construct of the language, there wasn't some optimization applied (not cloning a List of 10000 if not needed for example). I would be  surprised in this code if anything else than `ctry.types` was moved/cloned.

Reply:
Just to clarify: it is not wrong to use `match_sharedterm` here, it just adds a level of indirection that makes it a bit more difficult to read (in my opinion, of course). So I would advise to only use it when it brings a real benefit, like access to the `SharedTerm` later or a clear indication with a benchmark that we would need it there.

Reply:
> That's not really what match_sharedterm is used for. You would use it for cases where you still need access to the SharedTerm in case no match arm is selected, so that you avoid to duplicate a SharedTerm to another (maybe it's not very clear and I should clarify the documentation for the macro).

I don't understand why using or not the shared term in the `else` makes any difference. My understanding is that the goal of `matched_sharedterm` is to allow to take ownership of the underlying data *only if it matches a specific pattern*, instead of having to first perform a clone upfront that will be dropped without being used in some cases. Something that you can do transparently in a standard `match`, but not with something that is behind a `Rc`. If you need ownership in all branches, then on the contrary, you can call `into_owned` upfront as you will have to do it anyway, and use a standard match. It rather seems that `matched_sharederm` is useful only when you don't need to take ownership everywhere. Whether you reuse the original `Rc` or not when you don't take ownership doesn't seem very relevant to me.

Take the following case: a `RichTerm` is pointing to a `10000` elements `Term::List(ts)`, and it is a 2-Rc pointer. With the second solution, we will first clone blindly the underlying `Term` when calling the `into_owned()` function, and thus a `10000` elements array before matching. This won't match the pattern (that expects a `MetaValue`), and we will go to the `_` case: because we don't use the matched term, we will drop it right away. I'm afraid this won't be optimized away by `rustc/llvm`, as this is a pretty complex case. We do need the cloned term in the case it is a metavalue, which is only known at runtime. So it would need to push the `clone` inside all the branches, and then detects that one branches is unwraping an `Rc` and then dropping it without using it. That sounds like something [Koka](https://www.microsoft.com/en-us/research/uploads/prod/2020/11/perceus-tr-v1.pdf) would do, but from my (limited, let's be honest) knowledge of rust optimizations with respect to reference counting, I don't think it is anywhere near that level. But I may be wrong.

All of that being said, there's actually an important point to make, that I just realized only now: we are in the parser, and thus all `Rc` should be 1-counted, and no clone should ever take place when calling to `into_owned()`. Now, that does sound like a compelling case for having a simple `match` instead of using our own macro.

Reply:
In the end I changed it anyway, because I realized this must be a 1-Rc, and in this case the `match_shared` is indeed useless. If you do compare the results, I'm curious about it too :slightly_smiling_face: but to be honest this does look like reuse analysis with `Rc`s, that is described in the paper linked above. I would be very very surprised if Rust was able to do something like that.